### ISSUE TITLE
Problems v3.10.9

### ISSUE THREAD

**Issue by Romz24** [pawn-lang/compiler#367]
Why did you do this update? Now when compiling, I get a sea of warnings 239 and 214, thanks! =(

![kby7rijmzo4 1](https://user-images.githubusercontent.com/2458113/46287567-a2f6a800-c58b-11e8-930b-29d059c77af7.jpg)

**Comment by Y-Less** [comment]
Well why did you do that update then?

**Comment by Romz24** [comment]
I want to use the newest version but without these problems.

**Comment by Y-Less** [comment]
Then fix your code...

**Comment by YashasSamaga** [comment]
@Romz24 Do you think the warnings are wrong? Please check if you haven't yet. These warnings were added to point out mistakes in poorly written code which could potentially lead to bugs.

If you badly want to avoid these warnings and not correct your code, you can disable them. The best solution would be to actually fix your code. One fix could get rid of several warnings (maybe in hundreds).

**Comment by Romz24** [comment]
@YashasSamaga I have already decided to fix all the warnings, thanks. The main problem with the standard 'includes' that come with the SA-MP server.

**Comment by YashasSamaga** [comment]
@Romz24 That shouldn't have been happening. The feature does not run checks on native functions. Can you show the warnings which were triggered for the default includes?

**Comment by Romz24** [comment]
@YashasSamaga 

I get this error:
```
error 035: argument type mismatch (argument 4)
```

When using this code:
```
stock test(playerid, const nick[])
{
	ShowPlayerDialog(playerid, 1, DIALOG_STYLE_MSGBOX, nick, "text", "text", "text"); 
}
```

**Comment by Y-Less** [comment]
@Romz24 Get the fixed includes here: https://github.com/sampctl/samp-stdlib

**Comment by NexiusTailer** [comment]
@Y-Less But what if I do not want to update my default includes? There may be an update at any moment and then I'll face a bunch of new problems at once.

If the compiler can't work with default includes and lost compatibility with them, this means that it's only the compiler problems, not mine or default includes', and I also think this feature should be disabled (at least by default).

**Comment by Southclaws** [comment]
The samp-stdlib is kept up to date with the SA:MP includes, check out the readme for an explanation.

The problem here isn't with the compiler, it's the fact that the original default standard library includes are *wrong* - as in, they are not const-correct.

samp-stdlib exists to correct the issues that Kye has forgotten to fix over the years, such as missing functions, const-correctness, incorrect tags, etc. If you want to wait for Kye to fix these things, be my guest, but until then, samp-stdlib is a viable alternative that's community driven and actively maintained.

**Comment by NexiusTailer** [comment]
@Southclaws You've got me wrong, I want to use unmodified includes in order to avoid problems, when samp will be updated and my scripts will stop compiling without errors and warnings and I will have to re-patch new includes again and again. I don’t find it convenient, but moreover, I don’t think it was wise to introduce such a feature into the compiler when most people use the default includes. By default, first of all it should be maintained backward compatibility, and only those who want to see these warning should enable it themselves if they need.

**Comment by NexiusTailer** [comment]
Ok, I got your point. I didn't want to create problems for myself and users to the last to support both the default and custom compiler in my own scripts, suggesting that disabling these warning wouldn't create any obstacles to users who compile some scripts with the default includes and compiler and people who use custom compiler. In the first case there are some problems for now as script with logic for custom includes won't compile properly in the default compiler and vise-versa: some code intended for default includes & compiler will be failed on a custom one. So now any author of his own include (which adds some functions that calls default ones with non-const strings) has to either make some check for both compilers in every place of the script or make his code compileable for only one case. Very sad.

**Comment by Southclaws** [comment]
> I want to use unmodified includes in order to avoid problems
I'm not sure what you mean here, if you run into problems you can simply submit a PR, open an issue, etc. You can't do that with the default includes. There are no known issues with them because there are no real *changes* just additions of correct semantics and missing functions.

> I will have to re-patch new includes again and again
But you have to do that *anyway*, the only difference is the transport/medium (git pull vs download a zip and extract it)

I understand the confusion with "backwards compatbility" but remaining compatible with a mistake is the actual underlying debate here, and we have chosen to not introduce special-cases for old incorrect includes.

**Comment by Y-Less** [comment]
It really isn't hard to write for both at once, you just write your code to work with the more restrictive compiler and the more lenient one will still accept the code.

**Comment by NexiusTailer** [comment]
>I'm not sure what you mean here, if you run into problems you can simply submit a PR, open an issue, etc. You can't do that with the default includes. There are no known issues with them because there are no real changes just additions of correct semantics and missing functions.

Ok, here some cases which are already have created some stupid messes:

1. Having some code in include like:
```
public OnRconLoginAttempt(ip[], password[], success)
{
	if(!success) UniversalKick(INVALID_PLAYER_ID, ip, 1);
	return 1;
}

public OnVehicleMod(playerid, vehicleid, componentid)
{
	if(GetPlayerInterior(playerid) == 0)
	{
		UniversalKick(playerid, "", 0); // line 28
		return 0;
	}
	return 1;
}

static UniversalKick(playerid, ip_address[], type) //line 34
{
	if(type) BlockIpAddress(ip_address, 0);
	else Kick(playerid);
	return 0;
}
```

and using the custom compiler we will get 2 warnings:
```
myinc.inc(28) : warning 239: literal array/string passed to a non-const parameter
myinc.inc(34) : warning 214: possibly a "const" array argument was intended: "ip_address"
```

Ok. We just add missing "const" type to our "ip_address[]" string:
```
public OnRconLoginAttempt(ip[], password[], success)
{
	if(!success) UniversalKick(INVALID_PLAYER_ID, ip, 1);
	return 1;
}

public OnVehicleMod(playerid, vehicleid, componentid)
{
	if(GetPlayerInterior(playerid) == 0)
	{
		UniversalKick(playerid, "", 0);
		return 0;
	}
	return 1;
}

static UniversalKick(playerid, const ip_address[], type)
{
	if(type) BlockIpAddress(ip_address, 0); //line 36
	else Kick(playerid);
	return 0;
}
```

custom compiler (with modified includes) has no more complaints. But... what does a default one say to this?
```
myinc.inc(36) : error 035: argument type mismatch (argument 1)
```
Problems again...

So, different people with different compilers (or rather different includes, cause it depends on them whether the new compiler will compile without warnings/errors) will have problems. As a result, I **must** do stupid checks in almost every place where I substitute a "const" word (God forbid, such an argument will be passed to the default native). This is the inconvenience I was talking about.

**Comment by Y-Less** [comment]
Two more solutions:

1) Make the string literal not const:

```pawn
static gEmptyString[1] = "";
UniversalKick(playerid, gEmptyString, 0);
```

2) As I already said, use fixes.inc, which has had const-corrected natives for years.

**Comment by ziggi** [comment]
@NexiusTailer I write this bullshit for compatibility with unmodified include files:

https://github.com/Open-GTO/mdialog/blob/master/mdialog.inc#L68-L88
```Pawn
#if !defined FIX_const
	#define FIX_const 0
	#define MDIALOG_FIX_CLEAR
#endif

#if !FIX_const
	native MDialog_ShowPlayerDialog(playerid, dialogid, style, const caption[], const info[], const button1[], const button2[]) = ShowPlayerDialog;

	#if defined _ALS_ShowPlayerDialog
		#undef ShowPlayerDialog
	#else
		#define _ALS_ShowPlayerDialog
	#endif

	#define ShowPlayerDialog MDialog_ShowPlayerDialog

	#if defined MDIALOG_FIX_CLEAR
		#undef MDIALOG_FIX_CLEAR
		#undef FIX_const
	#endif
#endif
```

**Comment by NexiusTailer** [comment]
> @NexiusTailer I write this bullshit for compatibility with unmodified include files:
> 
> https://github.com/Open-GTO/mdialog/blob/master/mdialog.inc#L68-L88
> 
> ```pawn
> #if !defined FIX_const
> 	#define FIX_const 0
> 	#define MDIALOG_FIX_CLEAR
> #endif
> 
> #if !FIX_const
> 	native MDialog_ShowPlayerDialog(playerid, dialogid, style, const caption[], const info[], const button1[], const button2[]) = ShowPlayerDialog;
> 
> 	#if defined _ALS_ShowPlayerDialog
> 		#undef ShowPlayerDialog
> 	#else
> 		#define _ALS_ShowPlayerDialog
> 	#endif
> 
> 	#define ShowPlayerDialog MDialog_ShowPlayerDialog
> 
> 	#if defined MDIALOG_FIX_CLEAR
> 		#undef MDIALOG_FIX_CLEAR
> 		#undef FIX_const
> 	#endif
> #endif
> ```

Sorry for such a late answer, but this solution will break (ignore) the hook chain for that native if it was somewhere before your include. This is exactly the problem that some people faced in mdialog/nex-ac/[any other include with ALS hooks for ShowPlayerDialog] being used together, but I wrote about it here and now rather because this topic was touched again on in one of my repos.

**Comment by Y-Less** [comment]
That again seems like a problem with the out-of-date includes not being const-correct, not a problem with those trying to write const-correct code.  The correct solution there is to complain to the people that haven't updated their code yet, not the other way around.

**Comment by Y-Less** [comment]
@NexiusTailer I read the topic you referred to:

1. const-correct is, as the name implies, correct - that's why it was enabled, why it isn't disabled by default, and why the team isn't going to change it or apologise for it.

2. Supporting both compilers is very easy, I'm not sure why you're so stuck on that, because most other includes seem to manage it just fine.

3.  I just read your reply mostly as:

> I refuse to use third-party includes because they cause problems*.

> I have problems because I don't use third-party includes\*\*.

\* This is, of course, utterly false in 99% of cases, but still what you seem to believe.

\*\* samp-stdlib

**Comment by NexiusTailer** [comment]
> const-correct is, as the name implies, correct - that's why it was enabled, why it isn't disabled by default, and why the team isn't going to change it or apologise for it.

This can be correct for an idealist as much as possible, but there is still an "official" compiler and official samp includes, and having made the decision to break compatibility with them you just created a problem from scratch which has never been before.

> Supporting both compilers is very easy, I'm not sure why you're so stuck on that, because most other includes seem to manage it just fine.

Almost all the solutions suggested before are either problematic ways that have their own issues, or inappropriate ways for some cases I tried to use that in. The only really compact and suitable solution which was from the very beginning is to simply disable new warnings in the compiler if there is a new version of it. However, I have no idea what the team will radically change in future compiler versions, so I am still just watching a situation where people for no apparent reason decided to make us choose between the two, and I just made my own choice initially.

> 3. I just read your reply mostly as:
> 
> > I refuse to use third-party includes because they cause problems*.
> 
> > I have problems because I don't use third-party includes**.
> 
> * This is, of course, utterly false in 99% of cases, but still what you seem to believe.

As for this, you get the message a little bit wrong. I am not against using dependencies in gamemode or where their use is justified by the situation and there is a need for it, but I just see many cases where people add dependencies to their includes like just for fun and don't really understand why they should be used in a library that does fine without them and accordingly, the question arises about the justification of any of the dependencies as well justification to make user find and update it just to use one's single library (and if those libraries also have dependencies without real purpose, the user starting to face with some kind of dependency hell with all the consequences). I didn't say about sampctl users as this is still not all the samp community who use that.

**Comment by NexiusTailer** [comment]
> That again seems like a problem with the out-of-date includes not being const-correct, not a problem with those trying to write const-correct code.

Well, these includes are distributed from the official samp download section and are still part of the current samp server. I consider const-correct stdlib includes as a third-party solution and not an obligatory successor that everyone who first came to sa-mp.com and downloaded the archive must know about.

**Comment by Y-Less** [comment]
If your code is const-correct it will compile on either compiler.  The only difference is that one will warn you, the other won't.

**Comment by NexiusTailer** [comment]
But if I use default samp includes which is usually used with the default compiller, then there will be argument mismatch errors on it.

**Comment by Y-Less** [comment]
No there won't.  Argument mismatch means the size is wrong - you passed a variable to an array, or an array to a variable.  None of the data has changed size.  But I was wrong about never having warnings, I forgot you'll still get some on string literals passed to old natives on the new compiler.  But that means you have someone using the new compiler with the old includes, which is an odd combination.  Old compiler and old includes works, old compiler and new includes works, new compiler and new includes works, anything with fixes.inc works.  If you want to support weird combinations, where someone knew to update one part, but not another, then yes you'll probably have to do a little extra work (not a lot, it's still not hard):

```pawn
ShowPlayerDialog("Hello", other, params); // Fine with fixed (new) and broken (old) compiler/include combinations.
```

```pawn
static stock gHello[] = "hello";
ShowPlayerDialog(gHello, other, params); // Fine with all combinations (and identical compiled).
```

**Comment by NexiusTailer** [comment]
Argument mismatch also means that the arguments do not match their types, not just their number. I'll try to skip straight to the examples that create mentioned problems. Using the **default compiler and default includes** along [with this "correct" code](https://github.com/pawn-lang/compiler/issues/367#issuecomment-427297064) (in the second piece which have const) we get the following errors:
```myinc.inc(36) : error 035: argument type mismatch (argument 1)```

Solving this with such way:
```
static gEmptyString[1] = "";
UniversalKick(playerid, gEmptyString, 0);
```
We need to do it in every place we use that function, and even if we try to move that trick to the function itself, there will be still too many places we should change the code if we have a lot of similar native-with-string-args-powered functions. Just see one of weapon-config workarounds and how bulky it looks when many such cases accumulate in one place: https://i.imgur.com/jFxZXQO.png (and this is all just for the sake of "correcting" the already properly working code? allright). If we would have a thousand similar cases with different functions called in each case, we will have to do a thousand corrections for each one and this is not simple single change as you could say.

One more way [which were advised](https://github.com/pawn-lang/compiler/issues/367#issuecomment-434072210) and intended to manage const-problems in our hooks will just ignore some of the native "ShowPlayerDialog"'s hooks done before that (take ShowPlayerDialog as an example):
```
//========= include 1 ===========

stock firsthook_ShowPlayerDialog(playerid, dialogid, style, const caption[], const info[], const button1[], const button2[]) //Let's pretend here we have some another workaround or test it only on a new compiler/includes, "const" presence here not so important for the whole example
{
	printf("The first hook called");
	return ShowPlayerDialog(playerid, dialogid, style, caption, info, button1, button2);
}
#if defined _ALS_ShowPlayerDialog
	#undef ShowPlayerDialog
#else
	#define _ALS_ShowPlayerDialog
#endif
#define ShowPlayerDialog firsthook_ShowPlayerDialog

//========= include 2 ===========

native fix_ShowPlayerDialog(playerid, dialogid, style, const caption[], const info[], const button1[], const button2[]) = ShowPlayerDialog; //This is our workaround for now

#if defined _ALS_ShowPlayerDialog
	#undef ShowPlayerDialog
#else
	#define _ALS_ShowPlayerDialog
#endif
#define ShowPlayerDialog fix_ShowPlayerDialog

stock secondhook_ShowPlayerDialog(playerid, dialogid, style, const caption[], const info[], const button1[], const button2[])
{
	printf("The second hook called");
	return ShowPlayerDialog(playerid, dialogid, style, caption, info, button1, button2);
}
#if defined _ALS_ShowPlayerDialog
	#undef ShowPlayerDialog
#else
	#define _ALS_ShowPlayerDialog
#endif
#define ShowPlayerDialog secondhook_ShowPlayerDialog

//======== gamemode ==========

//Anywhere you want
ShowPlayerDialog(playerid, 0, DIALOG_STYLE_MSGBOX, "lol", "lol", "lol", "");
```
And if we add crashdetect plugin, we will see the reason of previous hookchain ignorance:
```
[15:56:16] [debug] Run time error 19: "File or function is not found"
[15:56:16] [debug]  firsthook_ShowPlayerDialog
[15:56:16] Script[gamemodes/nex-test.amx]: Run time error 19: "File or function is not found"
```

Cool, let's move to the next [advice of using fixes.inc](https://github.com/pawn-lang/compiler/issues/367#issuecomment-427313433). fixes.inc is a "lite" replacement for samp-stdlib and does not fix the problems themselves that occur in other includes, so its use can only help those people who tried to include one of the "const-correct" libraries, which cannot work with the default compiler/includes. That's not our case where we want make our code compatible with the two compilers and don't bother the end-user of our include to manage it.

One another solution for your own hooks of samp natives with the string args is to define [your own "MAYBE_CONST" macro](https://github.com/oscar-broman/samp-weapon-config/blob/master/weapon-config.inc#L96-L100) and use it instead of const in all problematic hooks. Worth admitting it rather a simple solution (and most importantly working one compared to the second way of native override). But it works only with samp-stdlib library used and not working, for example, with current version of fixes.inc if we use it to add const-correctness to our default samp includes. Sure it can be easily fixed by adding needed define in fixes.inc, but my words are aimed rather at exposing the problem that [such a solution](https://github.com/oscar-broman/samp-weapon-config/blob/master/weapon-config.inc#L96-L100) will be very specific and unreliable, as soon as the user implements any other ways of adding const-correctness and does not declare the corresponding signaling define.


At the same time that's the real actions, that I still think can happen on your part to remove unnecessary headaches from many people:

**Minimum set:**
* Make -z flag working for reverting const-correctness and keeping compatibility with the default compiler/includes again.
or
* Make defines like SAMP_CONST_CORRECT actually useful and working for compiler too: if it is not enabled, do some actions that will exclude all necessary default samp natives and callbacks without "const" from the const-correctness checks.

**Maximum set:**
* Just disable these new warnings by default. No problems who doesn't even care about it and no problems to enable it who really cares.

**Comment by Y-Less** [comment]
Half of those boil down to "if I use libraries that are out of date, they cause problems", to which the answer is just ask the author to update.  You're the first one I've ever encountered that flat-out refuses to do so, most other people understand why const-correctness is important, and accept that while it should have been done from the start, better late than never.  Your other solution is "go back to never" - just no.  Also, `-z` only reverts changes that actually break code, i.e. stop code compiling; not disable warnings.

Again, even if you don't want to use fixes.inc, at least look how it does things:

1. `#define ShowPlayerDialog( secondhook_ShowPlayerDialog(` - I've been advocating for this for years.  LONG before the new warnings were introduced, because it solves exactly the `native X() = Y;` problem you described.
2. It also defines all strings in the way I described, again since long before the new warnings, because it actually reduces memory usage with no overhead.  That image is not the way I said, nor the way fixes.inc does it, i.e. global static strings:

https://github.com/pawn-lang/sa-mp-fixes/blob/master/fixes.inc#L4408

Those are actually `const` (because they should be), but if you remain stubborn you can make them not `const`.  The way normal strings work is they are put in the same segment of memory as other `const` globals, i.e. defining the strings this way puts them in EXACTLY the same place as they would be were they normal inline string literals, and the code generated to access them is EXACTLY the same.  But the avantage of doing it this way are two-fold: if you drop `const` they're moved to the non-const region of memory and so can be passed to const-incorrect functions, and you can use the same string multiple times.  In the example you posted you had `"lol"` three times.  With every version of the compiler that will put `"lol"` in the AMX three times, with this version you have more control so can use the same copy every time.  Plus, there's no copying and uglyness as in the weapon-config version.  What is that @JustMichael?

But, again, most of your argument boils down to "not every include is const-correct", your solution is to thus say "stop writing good code", our solution is to say "then fix the includes".  We aren't going to sacrifice code quality because you, and others like you, can't be bothered to adapt to a change that has now been around for several years.  YSI, with like 50000 lines of code, managed to update; fixes.inc with 10,000 lines managed to adapt.  I'm sure Bobs 12 line ZMCD-Ex-v2 include can manage it if asked.

**Comment by NexiusTailer** [comment]
> Half of those boil down to "if I use libraries that are out of date, they cause problems", to which the answer is just ask the author to update.

You get the point as "libraries that is not updated", I get the point as "libraries that have compatibility with the default compiler/includes firstly". Every statement about "deprecated" or "out-of-date" is very easy to break against the fact that the "community" includes is a third-party library that is not distributed from the official site. Anyone can still download the current server archive with the up-to-date includes in it.

> Again, even if you don't want to use fixes.inc, at least look how it does things:

I didn't say I don't wanna use it, I said:
>[Using fixes.inc] not our case where we want make our code compatible with the two compilers and don't bother the end-user of our include to manage it.

Using fixes.inc just make us free of doing any compatibility with the default compiler/includes, but it start requiring fixes.inc as an obligatory dependency to make our include work without compilation problems on the "old" build tools.
 *** 
> But, again, most of your argument boils down to "not every include is const-correct", your solution is to thus say "stop writing good code", our solution is to say "then fix the includes".

No, its not my words nor solution and never was like that

> YSI, with like 50000 lines of code, managed to update;

And you don't even bother to test it on a default compiler, so that every N updates there are changes that cause compilation hangs, errors, or something like that, what I have witnessed many times. This is not a claim in your favor as the project is really huge, but it clearly shows that any compatibility with two compilers is not actually tested and made just for the record, not even saying about how much unnecessary code and crutches was written just to keep it working both for the two compilers.
